keps/sig-instrumentation/4827-component-statusz/README.md

```diff
@@ -0,0 +1,424 @@
+
+#### As an operator
+
+- I need to troubleshoot a performance issue and want to easily check for dependency conflicts or version mismatches without sifting through logs
+
+#### As support engineer
+
+- I want to access detailed build information to determine if a customer is running a known problematic version of a component
+
+### Risks and Mitigations
+
+1. **No sensitive data exposed**
+    
+  We will apply the  principle of least privilege by granting only the minimum necessary permissions for accessing /statusz, which is primarily a debugging tool for operators and maintainers.
+  /statusz will not be accessible through general discovery mechanisms like system:discovery. This prevents unintended exposure of potentially sensitive debugging information.
+  
+  We'll leverage Kubernetes RBAC to grant access to /statusz only to specific users and service accounts that require it for operational monitoring and debugging. For this, we will use the existing system:monitoring group ([ref](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L73)), ensuring that only authorized personnel can access this endpoint.
+  
+1. **Performance impact**
+    
+  We will take care to carefully consider the computational overhead associated with the statusz endpoint to avoid negatively impacting the server's overall performance.
+
+2. **Essential data only**
+    
+  We will prioritize inclusion of only essential diagnostic and monitoring data that aligns with the intended purpose of the z-page.
+
+3. **Premature Dependency on Unstable Format**
+
+  The alpha release will explicitly support plain text format only, making it clear that the output is not intended for parsing or automated monitoring. The feature will be secured behind a feature gate that is disabled by default, ensuring users opt-in consciously. Also, the endpoint will support specifying a version as a query parameter to facilitate future transitions to structured schema changes without breaking existing integrations, once the usage patterns and requirements are better understood
+
+## Design Details
+
+### Data Format and versioning
+
+Initially, the statusz page will exclusively support a plain text format for responses. In future, when we have established a structured schema for the response, we can support specifying a version param, like /statusz?version=2, to ensure future compatibility. Through this, we can evolve the endpoint and introduce changes to the response schema without impacting clients relying on previous versions, ensuring backward compatibility.
+
+### Authz and authn
+
+Access to the endpoint will be limited to members of the `system:monitoring group` ([ref](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L73)), which is how paths like /healthz, /livez, /readyz are restricted today. 
```

Comment:
for kubelet can you please add details about what the RBAC permissions will be for accessing this endpoint?

system:monitoring does not have the ability to call the kubelet's endpoints AFAIK. What subresource will allow a user access to kubelet's statsz endpoint?

Reply:
Hmm, I was assuming that `system:monitoring` group would suffice for kubelet as well, i.e. we could add the new endpoints to [this](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L227-L230) set of rules for system:monitoring role. 
cc @liggitt for his advice here.

Reply:
Does this endpoint expect to be programmatically queried? I was assuming no, which makes it feel different than the endpoints that are granted to `system:monitoring`.

Reply:
@richabanker - I think it would help to clarify what you will be adding. Can you add a markdown code section of what changes you will make to the `system:monitoring` ClusterRole.

Reply:
I don't think this will work. Here is how I tested this.

Create a serviceaccount
```
kubectl create serviceaccount hello --namespace default
```

Create a binding to system:monitoring ClusterRole
```
kubectl create clusterrolebinding --clusterrole=system:monitoring --serviceaccount default:hello helloissystemmonitor 
```

Call the kubelet metrics endpoint via apiserver proxy.
```
kubectl get --raw /api/v1/nodes/$(kubectl get nodes --no-headers | awk '{print $1}')/proxy/metrics --token $(kubectl create token hello)

Error from server (Forbidden): nodes "gke-test-default-pool-93fcb923-nt6x" is forbidden: User "system:serviceaccount:default:hello" cannot get resource "nodes/proxy" in API group "" at the cluster scope
```

Now lets try directly accessing the kubelet API of the node

Create a file called pod.yaml with these contents
```
apiVersion: v1
kind: Pod
metadata:
  name: ubuntu-sleeper
spec:
  serviceAccount: hello
  containers:
  - name: ubuntu
    image: ubuntu:latest # Use the latest Ubuntu image
    command: ["/bin/bash", "-c", "apt-get update && apt-get install -y curl jq && sleep infinity"] # Start a bash shell
    env:
    - name: NODE_IP
      valueFrom:
        fieldRef:
          fieldPath: status.hostIP
```

Create a pod

```
kubectl apply -f pod.yaml 
```

Exec into the pod

```
kubectl exec ubuntu-sleeper -it -- /bin/bash 
```

Run the following command inside the pod

```
root@ubuntu-sleeper:/# curl -k https://$NODE_IP:10250/metrics -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/
token)"
Forbidden (user=system:serviceaccount:default:hello, verb=get, resource=nodes, subresource=metrics)
```

Now lets check the permissions that hello has

```
kubectl auth can-i --list --as system:serviceaccount:default:hello              ok  test kube 
Warning: the list may be incomplete: webhook authorizer does not support user rule resolution
Resources                                       Non-Resource URLs                      Resource Names   Verbs
selfsubjectreviews.authentication.k8s.io        []                                     []               [create]
selfsubjectaccessreviews.authorization.k8s.io   []                                     []               [create]
selfsubjectrulesreviews.authorization.k8s.io    []                                     []               [create]
                                                [/.well-known/openid-configuration/]   []               [get]
                                                [/.well-known/openid-configuration]    []               [get]
                                                [/api/*]                               []               [get]
                                                [/api]                                 []               [get]
                                                [/apis/*]                              []               [get]
                                                [/apis]                                []               [get]
                                                [/healthz/*]                           []               [get]
                                                [/healthz]                             []               [get]
                                                [/healthz]                             []               [get]
                                                [/healthz]                             []               [get]
                                                [/livez/*]                             []               [get]
                                                [/livez]                               []               [get]
                                                [/livez]                               []               [get]
                                                [/livez]                               []               [get]
                                                [/metrics/slis]                        []               [get]
                                                [/metrics]                             []               [get]
                                                [/openapi/*]                           []               [get]
                                                [/openapi]                             []               [get]
                                                [/openid/v1/jwks/]                     []               [get]
                                                [/openid/v1/jwks]                      []               [get]
                                                [/readyz/*]                            []               [get]
                                                [/readyz]                              []               [get]
                                                [/readyz]                              []               [get]
                                                [/readyz]                              []               [get]
                                                [/version/]                            []               [get]
                                                [/version/]                            []               [get]
                                                [/version]                             []               [get]
                                                [/version]                             []               [get]
```

notice that it has /metrics permissions but that won't give it permissions for kubelet. For kubelet you will have to extend the rbac to something like nodes/statusz or something like that.

Reply:
I have added a markdown for changes that we can do to `system:monitoring` role to allow access to kubelet's /statusz. Does this look ok now @vinayakankugoyal ?

Reply:
Hmm... I am not sure about this.

It seems like system:monitoring did not previously have permissions to call kubelet API do we really want to grant this permission?

Reply:
I added these changes to `system:monitoring` as placeholder for now. We can have a final decision on this once @tallclair is available for comments. Also filed https://github.com/kubernetes/kubernetes/issues/127990 to track the discussion.

cc @SergeyKanzhelev